pkg/threescale/backend/helpers.go

```diff
@@ -112,3 +112,89 @@ func newApplication() *Application {
 		UnlimitedCounter: make(map[string]int),
 	}
 }
+
+func hasSurpassedDeadline(timestamp int64, granularity api.Period, deadline time.Time) bool {
+	switch granularity {
+	case api.Minute:
+		return time.Unix(timestamp, 0).Add(time.Minute).After(deadline)
+	case api.Hour:
+		return time.Unix(timestamp, 0).Add(time.Hour).After(deadline)
+	case api.Day:
+		return time.Unix(timestamp, 0).AddDate(0, 0, 1).After(deadline)
+	case api.Week:
+		// todo, here be dragons spawned by the forefathers
+		//    __        _
+		//  _/  \    _(\(o
+		//  /     \  /  _  ^^^o
+		//  /   !   \/  ! '!!!v'
+		//  !  !  \ _' ( \____
+		//  ! . \ _!\   \===^\)
+		//  \ \_!  / __!
+		//  \!   /    \
+		//  (\_      _/   _\ )
+		//  \ ^^--^^ __-^ /(__
+		//  ^^----^^    "^--v'
+		// time.Unix(timestamp, 0).AddDate(0, 0, 7).After(deadline)
+		return false
+	case api.Month:
+		return time.Unix(timestamp, 0).AddDate(0, 1, 0).After(deadline)
+	case api.Year:
+		return time.Unix(timestamp, 0).AddDate(1, 0, 0).After(deadline)
+	case api.Eternity:
+		return false
+	default:
+		return false
+	}
+}
+
+// statesAreComparable allows us to work with an applications counter and ensure that it makes sense to do calculations
+// on the provided index and metric. It also ensures we aren't duplicating code and checking for panics in the calling funcs
+// We can safely call this function with any index and know we wont panic. If the function returns false, the provided
+// args are unusable and we should handle in the caller appropriately
+// TODO - this may end up returning an error instead to provide more context if we need it for others purposes (eg logging)
+func statesAreComparable(metric string, index int, state, compareTo LimitCounter) bool {
+	localCounters, ok := state[metric]
+	if !ok || (len(localCounters)-1) < index {
+		return false
+	}
+
+	remoteCounter, ok := compareTo[metric]
+	if !ok || (len(remoteCounter)-1) < index {
+		return false
+	}
+
+	if localCounters[index].PeriodWindow.Period != remoteCounter[index].PeriodWindow.Period {
+		return false
+	}
+	return true
+}
+
+func getDifferenceAsMetrics(subtract, from LimitCounter) api.Metrics {
```

Comment:
This needs a general comment about what the func does - if it acts on LimitCounters then perhaps we'd want this as a method?

Reply:
well it would act likely be used to determine the difference between a snapshot and a new auth response after conversion I think